[JIT]  JIT should let people know we inferred an argument as a tensor

superwizard2019 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38527 [JIT]  JIT should let people know we inferred an argument as a tensor**

Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`

Differential Revision: [D21743876](https://our.internmc.facebook.com/intern/diff/D21743876)

Commit:
[JIT]  JIT should let people know we inferred an argument as a tensor

Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`

Reviewers:
meghanl,suo

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -661,6 +661,14 @@ struct CAFFE2_API TensorType : public Type {
+    return is_inferred_type_;
+  }
+
+  void setToInferredType(){
+    is_inferred_type_ = true;
+  }
```
I don't think we should expose this. It opens up the possibility that a `TensorType` instance will be created with `is_inferred_type` set to `false`, and then it will be set to `true` later in some other part of the code, perhaps by mistake. Given that many `Type` instances are shared using `TypePtrs` (for example, the instance returned by `TensorType::get()`), this can be difficult to debug.

We always know whether or not a `Tensor` type was inferred when creating/retrieving a pointer to a `TensorType` instance. Accordingly, I suggest creating a function like `TensorType::getInferred()` so that we can have a singleton for all of the uses of inferred `TensorType`.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -732,6 +740,8 @@ struct CAFFE2_API TensorType : public Type {
   // defined and undefined. However, no tensor type starts out with
   // `undefined_` set to `c10::nullopt`
   c10::optional<bool> undefined_;
+  // If a param doesn't have a type, default to "tensor"
```
Consider using a more descriptive comment here, like
```
// Represents whether or not this type was inferred.
```

SplitInfinity commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -174,6 +174,18 @@ static Value* tryMatchArgument(
           << arg.formatTypeMismatchMsg(value->type()->python_str());
 
+      if (auto pt = value->type()->cast<TensorType>()) {
+        if (pt->kind() == TypeKind::TensorType && pt->isInferredType()) {
```
If the cast to `TensorType` succeeded, then it must be true that `pt->kind() == TypeKind::TensorType`, so I think you can remove that check. It should be sufficient to check only that `pt->isInferredType()` is `true`. You can also put that on the same line as `auto pt = value...`.

SplitInfinity commented on torch/csrc/jit/frontend/schema_matching.cpp:
```diff
@@ -174,6 +174,18 @@ static Value* tryMatchArgument(
 
+      if (auto pt = value->type()->cast<TensorType>()) {
+        if (pt->kind() == TypeKind::TensorType && pt->isInferredType()) {
+          std::string inferred_type_hint;
+          inferred_type_hint = c10::str(
+              "Inferred the value for argument '",
+              arg.name(),
+              "' to be of type 'Tensor' ",
+              "because it was not annotated with an explicit type.\n");
+          ostream << inferred_type_hint;
```
This looks like it was copied from `function_schema.h`; can we refactor this somehow to avoid duplicating code?

SplitInfinity reviewed (commented):
Looks good, I have some comments.

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity commented:
There are a lot of changes in `aten/src/ATen/core/function_schema.h` and `aten/src/ATen/core/jit_type.h` that seem small and unrelated to your change (e.g. formatting changes, reordering of includes). Can you revert these so that it is easier for reviewers to focus on your changes?

superwizard2019 commented:
> There are a lot of changes in `aten/src/ATen/core/function_schema.h` and `aten/src/ATen/core/jit_type.h` that seem small and unrelated to your change (e.g. formatting changes, reordering of includes). Can you revert these so that it is easier for reviewers to focus on your changes?

Sure. I have been working on that :) I guess it is because I use the VScode Format tool by mistake.

SplitInfinity commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -59,14 +58,18 @@ struct Argument {
     return is_inferred_type_;
   }
 
+  static std::string formatInferredTypeHint(const std::string& name) {
+    return c10::str(
+        "Inferred '",
+        name,
+        "' to be of type 'Tensor' ",
+        "because it was not annotated with an explicit type.\n");
+  }
```
I made a mistake, can you move this back? The error message is slightly different for an inferred arg vs. an inferred *value* so we cannot refactor as easily as I thought. So sorry about this, it's my fault.

superwizard2019 replied:
OK sure

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

suo commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -661,6 +661,16 @@ struct CAFFE2_API TensorType : public Type {
+  }
+
+  static TensorTypePtr getInferred() {
+    auto pt = get();
```
get() returns a singleton instances of TensorType, so setting `is_inferred_type_` on it will flip it for every TensorType originating from this call.

Constructing a new, unique TensorType out of `get()` is a possible fix although may have performance implications.

SplitInfinity replied:
I was thinking of keeping another singleton here that's like what you get from `get()` but with the `is_inferred_type` set to `true`.

suo replied:
yeah that works too!

suo commented on torch/csrc/jit/frontend/script_type_parser.cpp:
```diff
@@ -271,7 +271,7 @@ std::vector<Argument> ScriptTypeParser::parseArgsFromDecl(
     if (!decl_arg.type().present()) {
       // If this param doesn't have a type, default to "tensor"
       is_inferred_type = true;
```
we are already populating `is_inferred_type` here, and passing that information to `Argument`. If we are adding it as metadata to TensorType, then we should remove the old code code that does the same thing.

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -36,6 +36,13 @@ struct Argument {
         kwarg_only_(kwarg_only),
         alias_info_(std::move(alias_info)),
         is_inferred_type_(is_inferred_type) {
```
Now that the information on whether or not the type is inferred is encoded in the `type_` member, let's remove this class member (`is_inferred_type_`) and constructor argument (`is_inferred_type`) altogether.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -542,6 +542,8 @@ struct CAFFE2_API TensorType : public Type {
       at::Device device,
       at::IntArrayRef sizes);
 
+  static TensorTypePtr createInferred();
```
This is only used in one place (`getInferred`), so I think we should not have a separate function for it for now. Let's inline it.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -661,6 +663,15 @@ struct CAFFE2_API TensorType : public Type {
   }
 
+  bool isInferredType() const {
+    return is_inferred_type_;
```
Let's fetch this value from `type_`.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -732,6 +743,8 @@ struct CAFFE2_API TensorType : public Type {
   // `undefined_` set to `c10::nullopt`
   c10::optional<bool> undefined_;
+  // Represents whether or not this type was inferred.
+  bool is_inferred_type_= false;
```
Nit: I think we can shorten the name of this member to `is_inferred_`.

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -56,7 +54,15 @@ struct Argument {
   bool is_inferred_type() const {
-    return is_inferred_type_;
+    bool is_inferred_type = false;
+    if (type_){
```
`type_` should never be `nullptr`. If it is, then something has gone wrong. I think we can replace this check with a `TORCH_INTERNAL_ASSERT` to eliminate one level of code nesting.

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -661,6 +661,17 @@ struct CAFFE2_API TensorType : public Type {
+    return is_inferred_;
+  }
+
+  static TensorTypePtr getInferred() {
```
This function will return a new instance every time it is called. I think we should use a singleton here to avoid that and reuse the same instance across calls. See `TensorType::get` for an example. As we discussed, you can modify or create and overload of `TensorType::create` if needed.

SplitInfinity reviewed (commented):
You also need to run `clang-format` to format your code correctly. The easiest way to do this is to run `tools/clang_format_all.py` and revert any changes to files you didn't touch.

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/jit_type.h:
```diff
@@ -661,6 +663,15 @@ struct CAFFE2_API TensorType : public Type {
+  }
+
+  static TensorTypePtr getInferred() {
+    static auto valueInferred = TensorType::create(true);
```
What do you think about adding a new argument to the existing `create` methods, and then changing this line to
```suggestion
    static auto valueInferred = TensorType::create(
      /*scalar_type=*/{}, /*device=*/{}, /*sizes=*/VaryingShape<ShapeSymbol>{}, /*stride=*/VaryingShape<Stride>{}, /*requires_grad=*/{}, /*undefined=*/false, /*is_inferred=*/true);
```
Then, you don't need an extra `TensorType::create(bool is_inferred)` function that does the same thing but sets `is_inferred` after creating the object.

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity commented on aten/src/ATen/core/type.cpp:
```diff
@@ -861,9 +861,12 @@ TensorTypePtr TensorType::create(
+    bool is_inferred) {
+    auto pt = TensorTypePtr(new TensorType(
       scalar_type, device, sizes, strides, requires_grad, undefined));
+    pt->is_inferred_ = is_inferred;
```
Nit: Consider doing this in the constructor for `TensorType`.

SplitInfinity reviewed (approved):
Looks good, thanks for working on this!

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

SplitInfinity reviewed (changes requested):
Hey, can you squish all of the commits into one before landing?

SplitInfinity reviewed (approved):
Never mind, disregard the last message!

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]

Commit:
Update on "[JIT]  JIT should let people know we inferred an argument as a tensor"


Summary:

This PR solves issue (#37200).
Error is encountered during IR generation while trying to resolve the call to sum.
Should let user know it inferred the value for argument 'dim' to be of type 'Tensor'
because it was not annotated with an explicit type.

Test Plan:

Add code to reprodue the issue (#37200)
`python test/test_jit.py TestJit.test_inferred_as_tensor`



[ghstack-poisoned]
